File: Raven.Client.Java/lib/src/main/java/raven/client/document/MultiDatabaseHiLoGenerator.java
package raven.client.document;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

import raven.abstractions.data.Constants;
import raven.client.connection.IDatabaseCommands;

public class MultiDatabaseHiLoGenerator {

  private final int capacity;
  private final ConcurrentMap<String, MultiTypeHiLoKeyGenerator> generators = new ConcurrentHashMap<String, MultiTypeHiLoKeyGenerator>();

  public MultiDatabaseHiLoGenerator(int capacity) {
    this.capacity = capacity;
  }


  public String generateDocumentKey(String dbName, IDatabaseCommands databaseCommands, DocumentConvention conventions, Object entity) {
    String key = dbName;
    if (key == null) {
      key = Constants.SYSTEM_DATABASE;
    }
    MultiTypeHiLoKeyGenerator generator = generators.putIfAbsent(dbName, new MultiTypeHiLoKeyGenerator(capacity));
    return generator.generateDocumentKey(databaseCommands, conventions, entity);
  }
}

File: Raven.Client.Lightweight/Document/MultiDatabaseHiLoGenerator.cs
// -----------------------------------------------------------------------
//  <copyright file="MultiDatabaseHiLoGenerator.cs" company="Hibernating Rhinos LTD">
//      Copyright (c) Hibernating Rhinos LTD. All rights reserved.
//  </copyright>
// -----------------------------------------------------------------------
using System.Collections.Concurrent;
using Raven.Abstractions.Data;
using Raven.Client.Connection;

namespace Raven.Client.Document
{
	public class MultiDatabaseHiLoGenerator
	{
		private readonly int capacity;

		private readonly ConcurrentDictionary<string, MultiTypeHiLoKeyGenerator> generators =
			new ConcurrentDictionary<string, MultiTypeHiLoKeyGenerator>();

		public MultiDatabaseHiLoGenerator(int capacity)
		{
			this.capacity = capacity;
		}

		public string GenerateDocumentKey(string dbName, IDatabaseCommands databaseCommands, DocumentConvention conventions, object entity)
		{
			var multiTypeHiLoKeyGenerator = generators.GetOrAdd(dbName ?? Constants.SystemDatabase, s => new MultiTypeHiLoKeyGenerator(capacity));
			return multiTypeHiLoKeyGenerator.GenerateDocumentKey(databaseCommands, conventions, entity);
		}
	}
}